### PR TITLE
feat: build image by platform

### PR DESCRIPTION
**Pull request by n063h**
refer to #1580

# Before
Hardcoded linux/amd64 platform into code, so envd builds image on linux/amd64 platform by default.

# After
support --platform flag to build image on target platform.

### CODE REVIEW

**Comment by pingsutw** [comment]
@n063h Thanks for working on this. Any update for it. we need muti-arch image. I can help for testing

**Comment by n063h** [comment]
> @n063h Thanks for working on this. Any update for it. we need muti-arch image. I can help for testing

@pingsutw  Thanks for reaching out! There's a small hiccup with the dependency "horust" image that I'm working on fixing. I'll get it sorted out soon.

**Comment by n063h** [comment]
@pingsutw Hi, sorry for the wait, I think this branch is ready for your test now. You can run `envd build --platform linux/amd64,linux/arm64`  to build multi-platform images.
Since I have no machine on Non-x86 platforms, I tested it with qemu on linux/arm64 which worked well. 
Please note: horust now provides only linux/amd64,linux/arm64 images, so the images built from envd also work just for the two platforms.

**Comment by kemingy** [comment]
Is it ready to review?

BTW the dev env only work for aarch64/amd64, but the non-dev env should work for more platforms.

**Comment by n063h** [comment]
> Is it ready to review?
> 
> BTW the dev env only work for aarch64/amd64, but the non-dev env should work for more platforms.

yes. 
And the default target-platform is "linux/amd64" no matter what the build-platform is, which may require explicitly flag for arch arm64 for macos users.

**Review by gaocegege — APPROVED** [review]
/cc @kemingy

File: pkg/app/build.go [hunk 1]
```diff
@@ -106,12 +112,23 @@ func build(clicontext *cli.Context) error {
...
 
-	builder, err := buildutil.GetBuilder(clicontext, opt)
-	if err != nil {
-		return err
-	}
-	if err = buildutil.InterpretEnvdDef(builder); err != nil {
-		return err
+	platforms := strings.Split(opt.Platform, ",")
+	for _, platform := range platforms {
+		o, p := opt, platform
```
Comment (kemingy) [inline pkg/app/build.go:117]: May not need to copy the platform here?

File: pkg/app/build.go [hunk 2]
```diff
@@ -106,12 +112,23 @@ func build(clicontext *cli.Context) error {
...
+		o.Platform = p
+		if len(platforms) > 1 {
+			o.Tag += "-" + strings.Replace(p, "/", "-", 1)
+		}
```
Comment (kemingy) [inline pkg/app/build.go:121]: What does this do? Can you add some comments?

File: pkg/app/build.go [hunk 3]
```diff
@@ -89,6 +90,11 @@ To build and push the image to a registry:
 			Usage:   "Import the cache (e.g. type=registry,ref=<image>)",
 			Aliases: []string{"ic"},
 		},
+		&cli.StringFlag{
+			Name:        "platform",
+			Usage:       "Specify the target platforms for the build output (for example, windows/amd64 or linux/amd64,darwin/arm64)",
+			DefaultText: "linux/amd64",
```
Comment (kemingy) [inline pkg/app/build.go:96]: I wonder if we should use the local host platform as the default one.

File: pkg/app/build.go (hunk 2 above)
Comment (n063h) [reply pkg/app/build.go:121]: build images with same tag could cause image overwriting, so here I add an platform suffix to differentiate them.

File: pkg/app/build.go (hunk 3 above)
Comment (n063h) [reply pkg/app/build.go:96]: Ok, I will use the GOOS/GOARCH as the default one

File: pkg/app/build.go (hunk 2 above)
Comment (kemingy) [reply pkg/app/build.go:121]: This needs to be documented in CLI help. We'd better create a new issue to track it.

**Comment by kemingy** [comment]
LGTM. Please resolve the conflicts.

**Review by pingsutw — COMMENTED** [review]
could we also update the platform for [NewImageSource](https://github.com/tensorchord/envd/blob/fd45847e4ebec79bf644d291fa4d2420ea032373/pkg/lang/ir/util.go#L35)
```go
sys := types.SystemContext{}
if platform != nil {
  sys.ArchitectureChoice = platform.Architecture
  sys.OSChoice = platform.OS
}
src, err := ref.NewImageSource(ctx, &sys)
```

**Comment by pingsutw** [comment]
It will pull the base image. I'm using envd in my macbook. It still tries to pull darwin/amd64 base image, although I set the platform to linux/amd64, so we also need to update the platform for NewImageSource.
https://github.com/tensorchord/envd/pull/1587

**Comment by n063h** [comment]
> It will pull the base image. I'm using envd in my macbook. It still tries to pull darwin/amd64 base image, although I set the platform to linux/amd64, so we also need to update the platform for NewImageSource. #1587

I got you, the update has been pushed.

**Review by pingsutw — APPROVED** [review]
LGTM, thank you

**Comment by pingsutw** [comment]
btw, @kemingy could you help me cut a release after it merges? I need to bump envd version in flytekit.

**Review by kemingy — APPROVED** [review]

**Comment by kemingy** [comment]
> btw, @kemingy could you help me cut a release after it merges? I need to bump envd version in flytekit.

https://github.com/tensorchord/envd/releases/tag/v0.3.22 is released

**Pull request merged**